File: ts-bhcs/src/core/engine.ts
/**
 * Homeostatic Engine - Core Regulation Logic
 * 
 * Implements the fundamental homeostatic equation:
 * error = target - current
 * adjustment = learning_rate * error
 * new_state = current + adjustment
 */

export interface ZoneState {
    id: number;
    activity: number;
    target: number;
    state: 'CALM' | 'OVERSTIMULATED' | 'EMERGENT' | 'CRITICAL';
    lastUpdate: number;
}

export interface HomeostaticConfig {
    targetCalmness: number;
    learningRate: number;
    zones: number;
    updateInterval: number;
}

export class HomeostaticEngine {
    private zones: Map<number, ZoneState>;
    private config: HomeostaticConfig;
    private isRunning: boolean = false;

    constructor(config?: Partial<HomeostaticConfig>) {
        this.config = {
            targetCalmness: 0.5,
            learningRate: 0.02,
            zones: 5,
            updateInterval: 1000,
            ...config
        };

        this.zones = new Map();
        this.initializeZones();
    }

    /**
     * Initialize all zones with default states
     */
    private initializeZones(): void {
        for (let i = 0; i < this.config.zones; i++) {
            const zone: ZoneState = {
                id: i,
                activity: Math.random() * 0.6 + 0.2, // Random initial activity
                target: this.config.targetCalmness,
                state: this.calculateState(0.5),
                lastUpdate: Date.now()
            };
            this.zones.set(i, zone);
        }
    }

    /**
     * Initialize the engine
     */
    public async initialize(): Promise<void> {
        console.log('🦀 Homeostatic Engine: Zones initialized');
        this.isRunning = true;
    }

    /**
     * Main update loop - applies homeostatic regulation
     */
    public async update(): Promise<void> {
        if (!this.isRunning) return;

        const now = Date.now();
        
        for (const [id, zone] of this.zones) {
            // Apply homeostatic update equation
            const error = zone.target - zone.activity;
            const adjustment = this.config.learningRate * error;
            const newActivity = Math.max(0, Math.min(1, zone.activity + adjustment));
            
            // Update zone state
            zone.activity = newActivity;
            zone.state = this.calculateState(newActivity);
            zone.lastUpdate = now;
            
            this.zones.set(id, zone);
        }
    }

    /**
     * Calculate zone state based on activity level
     */
    private calculateState(activity: number): ZoneState['state'] {
        if (activity < 0.4) return 'CALM';
        if (activity < 0.7) return 'OVERSTIMULATED';
        if (activity < 0.9) return 'EMERGENT';
        return 'CRITICAL';
    }

    /**
     * Apply external influence to a zone
     */
    public applyInfluence(zoneId: number, influence: number): void {
        const zone = this.zones.get(zoneId);
        if (!zone) return;

        zone.activity = Math.max(0, Math.min(1, zone.activity + influence));
        zone.state = this.calculateState(zone.activity);
        zone.lastUpdate = Date.now();
        
        this.zones.set(zoneId, zone);
        console.log(`🎛️ Applied influence ${influence.toFixed(3)} to Zone ${zoneId}`);
    }

    /**
     * Get current state of all zones
     */
    public getZoneStates(): ZoneState[] {
        return Array.from(this.zones.values());
    }

    /**
     * Get specific zone state
     */
    public getZoneState(zoneId: number): ZoneState | undefined {
        return this.zones.get(zoneId);
    }

    /**
     * Get system-wide metrics
     */
    public getSystemMetrics() {
        const zones = Array.from(this.zones.values());
        const avgActivity = zones.reduce((sum, zone) => sum + zone.activity, 0) / zones.length;
        const stateDistribution = zones.reduce((acc, zone) => {
            acc[zone.state] = (acc[zone.state] || 0) + 1;
            return acc;
        }, {} as Record<string, number>);

        return {
            averageActivity: avgActivity,
            totalZones: zones.length,
            stateDistribution,
            homeostaticBalance: Math.abs(avgActivity - this.config.targetCalmness),
            timestamp: Date.now()
        };
    }

    /**
     * Reset engine to initial state
     */
    public reset(): void {
        this.zones.clear();
        this.initializeZones();
        console.log('🔄 Homeostatic Engine reset to initial state');
    }

    /**
     * Stop the engine
     */
    public stop(): void {
        this.isRunning = false;
        console.log('🛑 Homeostatic Engine stopped');
    }
}

File: rust-core/src/engine.rs
//! Homeostatic Engine Implementation
//! 
//! Core regulation logic for BHCS deterministic control

use serde::{Deserialize, Serialize};
use crate::zone::Zone;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HomeostaticConfig {
    pub target_calmness: f64,
    pub learning_rate: f64,
    pub zones: usize,
}

impl Default for HomeostaticConfig {
    fn default() -> Self {
        Self {
            target_calmness: 0.5,
            learning_rate: 0.02,
            zones: 5,
        }
    }
}

#[derive(Debug, Clone)]
pub struct HomeostaticEngine {
    zones: Vec<Zone>,
    config: HomeostaticConfig,
}

impl HomeostaticEngine {
    pub fn new(config: HomeostaticConfig) -> Self {
        let mut zones = Vec::new();
        for i in 0..config.zones {
            zones.push(Zone::new(i));
        }
        
        Self { zones, config }
    }

    pub fn update(&mut self) {
        for zone in &mut self.zones {
            // Apply homeostatic update equation
            let error = self.config.target_calmness - zone.activity();
            let adjustment = self.config.learning_rate * error;
            zone.apply_adjustment(adjustment);
        }
    }

    pub fn apply_influence(&mut self, zone_id: usize, influence: f64) {
        if let Some(zone) = self.zones.get_mut(zone_id) {
            zone.apply_influence(influence);
        }
    }

    pub fn get_zones(&self) -> &[Zone] {
        &self.zones
    }

    pub fn get_zone(&self, zone_id: usize) -> Option<&Zone> {
        self.zones.get(zone_id)
    }

    pub fn get_system_metrics(&self) -> SystemMetrics {
        let avg_activity = self.zones.iter()
            .map(|z| z.activity())
            .sum::<f64>() / self.zones.len() as f64;

        let homeostatic_balance = (avg_activity - self.config.target_calmness).abs();

        SystemMetrics {
            average_activity: avg_activity,
            total_zones: self.zones.len(),
            homeostatic_balance,
            timestamp: chrono::Utc::now().timestamp(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemMetrics {
    pub average_activity: f64,
    pub total_zones: usize,
    pub homeostatic_balance: f64,
    pub timestamp: i64,
}
